### PR TITLE
Add Gnosis Safe connector

### PR DESCRIPTION
**Pull request by chnejohnson**
## Issues
- #52 
- #89 

## Reference
- https://github.com/safe-global/safe-apps-sdk/tree/master/packages/safe-apps-wagmi
- https://github.com/sushiswap/sushiswap/pull/217/files
- https://github.com/re2005/vue-gnosis-safe
- https://ethereum.stackexchange.com/questions/120269/auto-safeapp-detection/120339#120339
- https://stackoverflow.com/questions/67259947/access-to-web-manifest-blocked-by-cors-policy

## Test
1. `yarn build:demo` and `yarn serve:demo`
2. check `http://localhost:4173` in the [Safe app](https://gnosis-safe.io/app) (Safe App > Apps > Add custom app)

### CODE REVIEW

**Comment by chnejohnson** [comment]
@re2005 Please help me review the code and feel free to give me any advice to improve it.

File: src/connectors/safe.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+import { normalizeChainId } from '../utils'
+import { Connector } from './connector'
+import {
+  ConnectorNotFoundError,
+  ProviderNotFoundError,
+  ProviderRpcError,
+} from './errors'
+
+const __IS_SERVER__ = typeof window === 'undefined'
+const __IS_IFRAME__ = !__IS_SERVER__ && window?.parent !== window
```
Comment (re2005) [inline src/connectors/safe.ts:16]: Would be nice to export these flags, so users can use it on their apps to conditionally include the connector or NOT.

Like inside GnosisSafe, the `only` connector you should use if the Safe, the other won't work.

For example in my dapo I'm doing this:

```ts
const isServer = typeof window === 'undefined';
const isIframe = !isServer && window?.parent !== window;

const connectors: any = [
    new MetaMaskConnector({
        appUrl: window.location.href,
    }),
    new WalletConnectConnector({
        qrcode: true,
        rpc: Object.values(supportedChains).reduce((acc, cur) => ({ ...acc, [cur.chainId]: cur.rpcUrls[0] }), {}),
    }),
];

if (isIframe) {
    connectors.push(new SafeConnector({}));
}
```

Maybe is nice if I can import from 'vue-dapp`

```
import {isGnosisSafe} from 'vue-dapp'
```

I believe users shouldn't show other connectors in here, only `GnosisSafe`

Screenshot 2022-10-29 at 12 27 51

**Comment by re2005** [comment]
Somehow the Gnosis is missing the manifest. Perhaps it should be linked on the `index.html` of the demo.

missing-manifest

Could try adding it ?

**Comment by re2005** [comment]
Overall it's looking good. Very nice feature to have.

Later this weekend I'll test within my own dapps and post here more feedbacks.

Nice work @chnejohnson!

**Comment by chnejohnson** [comment]
Thanks @re2005 for your feedback!

In the commit [59d5e26](https://github.com/chnejohnson/vue-dapp/pull/94/commits/59d5e2674eae067d76e94b77046881ade9720037), `isNotSafeApp` only check whether it's in the iframe to detect Safe app. So it's possible that the app is in the iframe but not inside Gnosis Safe app.

```js
if (!isNotSafeApp()) {
  const safe = new SafeConnector()
  connectors = [safe, connectors[0]]
}
```

A strict way is to use `safe.isSafeApp()` to detect the Safe app.
But to check whether it's exactly a safe app, it has to call a Promise, and it will lead to some bugs happening in the `Board.vue`'s `props.connectors`
```ts
const safe = new SafeConnector()
const isSafeApp = await safe.isSafeApp()
if (isSafeApp) {
  connectors = [safe, connectors[0]]
}
```
At the moment, I haven't found a good solution to solve this. So just implement the first one for the feature of conditional connector inclusion.

**Comment by chnejohnson** [comment]
Besides, I added the manifest link but it seems that it still cannot be detected on Netlify preview site. 
Maybe the formal site can be detected. 

截圖 2022-11-02 上午11 24 12

截圖 2022-11-02 上午11 24 29

**Comment by re2005** [comment]
Hi @chnejohnson thanks for the update. I've tested not only the demo, but also on my application. It all looks good, including the metadata.

I believe you can move on and merge it.

Great job!

**Comment by chnejohnson** [comment]
@re2005 glad to hear that, thanks for your review!

**Comment by chnejohnson** [comment]
> Thanks @re2005 for your feedback!
> 
> In the commit [59d5e26](https://github.com/chnejohnson/vue-dapp/pull/94/commits/59d5e2674eae067d76e94b77046881ade9720037), `isNotSafeApp` only check whether it's in the iframe to detect Safe app. So it's possible that the app is in the iframe but not inside Gnosis Safe app.
> 
> ```js
> if (!isNotSafeApp()) {
>   const safe = new SafeConnector()
>   connectors = [safe, connectors[0]]
> }
> ```
> 
> A strict way is to use `safe.isSafeApp()` to detect the Safe app. But to check whether it's exactly a safe app, it has to call a Promise, and it will lead to some bugs happening in the `Board.vue`'s `props.connectors`
> 
> ```ts
> const safe = new SafeConnector()
> const isSafeApp = await safe.isSafeApp()
> if (isSafeApp) {
>   connectors = [safe, connectors[0]]
> }
> ```
> 
> At the moment, I haven't found a good solution to solve this. So just implement the first one for the feature of conditional connector inclusion.

I found a solution to detect the Safe app in a strict way.
```ts
const connectorsCreated = ref(false)

onMounted(async () => {
  const safe = new SafeConnector()
  if (await safe.isSafeApp()) {
    connectors = [safe]
  }
  connectorsCreated.value = true
})
```

```vue
 <vd-board v-if="connectorsCreated" :connectors="connectors" dark>
```

**Pull request merged**